Add new `customer_choice` module

drbenvincent opened the pull request:
This PR will add new functionality to pymc-marketing. Namely it will allow causal analysis of product incrementality/cannibalization.

In this initial PR we will focus on:
a) adding docs to introduce the idea behind incrementality/cannibalization analysis
b) in terms of modeling we will start with a multivariate interrupted time series model. This will include a notebook to illustrate its use, as well as docs in the guide section to convey the basic workings.

Later PR's will add additional modeling options, such as nested discrete choice models.

## TODO
- [X] Finish "Introduction to product incrementality" docs page
- [x] Finish MV-ITS model explanation
- [x] Add tests
- [x] Finish how-to notebook demonstrating use
  - [X] Add module code
  - [x] Add Scenario 3 where we operate on product level data
  - [x] Add scenario where market is not saturated + update relevant docs + tests
- [x] Add a QuickStart on the Getting Started docs page
  - [x] add python code block for QuickStart 
- [x] Rename module to `consumer-choice`. Update docs and imports and re-run notebooks etc.
- [x] Fix typo "isoluate" -> "isolate"
- [ ] Update docs homepage and `README.md`

## Known limitations
At the moment the MVITS model is very simple. It only has an intercept term, embodying the assumption that sales are relatively static over time. Clearly there are many ways to expand this model, so the proposal is that we get this in as a first attempt and stepwise add features over time. We can be lead by feature requests (if there are any), but this establishes an MVP which can be built upon.

## Checklist
- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Modules affected
This PR will create a new module and will not affect the MMM or CLV modules.

## Type of change
- [X] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):


<!-- readthedocs-preview pymc-marketing start -->
----
📚 Documentation preview 📚: https://pymc-marketing--803.org.readthedocs.build/en/803/

<!-- readthedocs-preview pymc-marketing end -->

juanitorduz commented on tests/incrementality/test_incrementality.py:
```diff
@@ -0,0 +1,100 @@
+sample_kwargs = {"tune": 100, "draws": 100}
+
+
+def generate_data(
```
This can be a fixture

drbenvincent replied:
done

```python
@pytest.fixture
def saturated_data_fixture():
    return generate_saturated_data(**scenario_saturated)
```

juanitorduz commented on tests/incrementality/test_incrementality.py:
```diff
@@ -0,0 +1,100 @@
+
+
+def test_MVITS():
+    data = generate_data(**scenario)
+    result = MVITS(
+        data,
+        treatment_time=scenario["treatment_time"],
+        background_sales=["competitor", "own"],
+        innovation_sales="new",
+        rng=rng,
+        sample_kwargs=sample_kwargs,
+    )
+    assert isinstance(result, MVITS)
+
+    ax = result.plot_fit()
+    assert isinstance(ax, plt.Axes)
+
+    ax = result.plot_counterfactual()
+    assert isinstance(ax, plt.Axes)
+
+    ax = result.plot_causal_impact()
+    assert isinstance(ax, plt.Axes)
+
+    result.plot_causal_impact(type="sales")
+    assert isinstance(ax, plt.Axes)
+
+    result.plot_causal_impact(type="market_share")
+    assert isinstance(ax, plt.Axes)
```
You can parametrize this with a fixture as in 

https://github.com/pymc-labs/pymc-marketing/blob/05abba75c9c90ecf064c398722c4ecb9e93219e1/tests/mmm/test_plotting.py#L126-L153

So that you ne not need to repeat the assert statement many times

drbenvincent replied:
I might try to push back on this one. Right now I have 2 tests which have this plot/assert pattern (`test_MVITS_saturated` and `test_MVITS_unsaturated`)

So the thing is that `test_MVITS_unsaturated` is already parameterised to run through multiple scenarios. I was able to add 2 parameterisations, but then it took much longer because it was fitting a model then testing one plot function. So then I'd have to do the model fitting in a fixture as well and it starts getting high non-obvious to read.

I'd also say that while
```python
ax = result.plot_fit()
assert isinstance(ax, plt.Axes)

ax = result.plot_counterfactual()
assert isinstance(ax, plt.Axes)

result.plot_causal_impact_sales()
assert isinstance(ax, plt.Axes)

result.plot_causal_impact_market_share()
assert isinstance(ax, plt.Axes)
```
has some redundancy, I think it is supremely readable.

Happy to consider doing something if you really don't like this approach, but otherwise I'm tempted to leave it as it is.

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,298 @@
+        )
+        return ax
+
+    def plot_causal_impact(self, type="sales"):
```
Will type ever be more then these two values? If not, breaking out into two methods might allow for easier discover

Also not fan of name type as it is a Python keyword

drbenvincent replied:
Thanks. Have split up into two different plot methods

drbenvincent commented:
Thanks for the comments @juanitorduz + @wd60622. I'll get on those soon. Should be able to mark as ready for review soon.

drbenvincent commented:
Failing test `FAILED tests/test_prior.py::test_constrain` is nothing to do with me ;)

drbenvincent commented:
I'm not 100% sure why the code coverage is failing. I've got integration tests in there, so assumed that would do a good enough job 🤷🏻‍♂️

drbenvincent commented:
Tagging @aseyboldt in case you want to browse the docs https://pymc-marketing--803.org.readthedocs.build/en/803/ and see if there are any refinements that could be made.

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+        self.sample_kwargs = sample_kwargs if sample_kwargs is not None else {}
+        self.market_saturated = market_saturated
+
+        # build the model
+        self.model = self.build_model(
+            self.data[self.background_sales],
+            self.data[self.innovation_sales],
+            self.market_saturated,
+            treatment_time=self.treatment_time,
+        )
+
+        # sample from prior, posterior, posterior predictive
+        with self.model:
+            self.idata = pm.sample_prior_predictive(random_seed=self.rng)
+            self.idata.extend(pm.sample(**self.sample_kwargs, random_seed=self.rng))
+            self.idata.extend(
+                pm.sample_posterior_predictive(
+                    self.idata,
+                    var_names=["mu", "y"],
+                    random_seed=self.rng,
+                )
+            )
+
+        # Calculate the counterfactual background sales, if the new product had not been introduced
+        zero_sales = np.zeros(self.data[self.innovation_sales].shape, dtype=np.int32)
+        self.counterfactual_model = pm.do(self.model, {"innovation_sales": zero_sales})
+        with self.counterfactual_model:
+            self.idata_counterfactual = pm.sample_posterior_predictive(
+                self.idata, var_names=["mu", "y"], random_seed=self.rng
+            )
+
+        return
```
Think this should be separate from init. This might be sample_prior_preditive, fit, sample_posterior_predictive methods

drbenvincent replied:
Sure. I've split into smaller methods outside of init.

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+        zero_sales = np.zeros(self.data[self.innovation_sales].shape, dtype=np.int32)
+        self.counterfactual_model = pm.do(self.model, {"innovation_sales": zero_sales})
+        with self.counterfactual_model:
+            self.idata_counterfactual = pm.sample_posterior_predictive(
+                self.idata, var_names=["mu", "y"], random_seed=self.rng
+            )
```
Save and load methods would require storing two idatas. Think this could be combined with the other inference data? Might have to rename the keys or just take posterior_predictive group

drbenvincent replied:
Good idea. Done

wd60622 reviewed (commented):
Some more thoughts

lucianopaz commented on tests/incrementality/__init__.py:
You should follow the same name convention as in the package. Use either `product_incrementality` or `incrementality` everywhere.

drbenvincent replied:
Good point. I've updated it all to `product_incrementality`. Note that the exact name may change based on discussions we have about how we 'sell' the MV-ITS and DCM stuff. But that's for a later discussion.

lucianopaz commented on tests/incrementality/test_incrementality.py:
```diff
@@ -0,0 +1,130 @@
+sample_kwargs = {"tune": 100, "draws": 100}
+
+
+@pytest.fixture
```
Use an explicit scope for this fixture. I don’t know what the default is, but I think it’s better to be explicit and set it to “module”

lucianopaz commented on tests/incrementality/test_incrementality.py:
```diff
@@ -0,0 +1,130 @@
+    return generate_saturated_data(**scenario_saturated)
+
+
+def test_plot_data(saturated_data_fixture):
```
Maybe you could add mocks to the arviz and matplotlib functions to actually test that the function is performing the calls you want with the arguments you expect?

lucianopaz commented on tests/incrementality/test_incrementality.py:
```diff
@@ -0,0 +1,130 @@
+    ax = result.plot_counterfactual()
+    assert isinstance(ax, plt.Axes)
+
+    result.plot_causal_impact_sales()
```
You are not collecting the result from this call. I imagine you want to start this line with `ax =`.

drbenvincent replied:
Fixed

lucianopaz commented on tests/incrementality/test_incrementality.py:
```diff
@@ -0,0 +1,130 @@
+    result.plot_causal_impact_sales()
+    assert isinstance(ax, plt.Axes)
+
+    result.plot_causal_impact_market_share()
```
Same as above

drbenvincent replied:
Fixed

lucianopaz commented on tests/incrementality/test_incrementality.py:
```diff
@@ -0,0 +1,130 @@
+    ax = result.plot_counterfactual()
+    assert isinstance(ax, plt.Axes)
+
+    result.plot_causal_impact_sales()
```
Same comment about `ax =`

drbenvincent replied:
Fixed

lucianopaz commented on tests/incrementality/test_incrementality.py:
```diff
@@ -0,0 +1,130 @@
+    result.plot_causal_impact_sales()
+    assert isinstance(ax, plt.Axes)
+
+    result.plot_causal_impact_market_share()
```
Same as above

drbenvincent replied:
Fixed

lucianopaz commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+        """Return a PyMC model for a multivariate interrupted time series analysis."""
+
+        if not background_sales.index.equals(innovation_sales.index):
+            raise ValueError(
```
Do you want to add a test for this? If not, just add a no-cover statement so codecov doesn’t complain

lucianopaz commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+            # expectation
+            mu = pm.Deterministic(
+                "mu",
+                intercept[None, :] - innovation_sales[:, None] * beta[None, :],
```
What are the innovation sales and the background sales supposed to mean? Do you expect this to be a general enough relationship? Is the term innovation supposed to distinguish between a set of products that you consider the treatment (or target, I don’t remember the proper name) for the interrupted time series, and another set of products that aren’t?

lucianopaz reviewed (changes requested):
@drbenvincent, I went through the code but I’ll have to look through the notebooks later.

drbenvincent commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+        # Note: if we compare "mu" then we are comparing the expected sales,
+        # if we compare "y" then we are comparing the actual sales
+        if variable not in ["mu", "y"]:
+            raise ValueError(f"variable must be either 'mu' or 'y', not {variable}")
```
added `# pragma: no cover`

drbenvincent commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+        fig, ax = plt.subplots()
+
+        if variable not in ["mu", "y"]:
+            raise ValueError(f"variable must be either 'mu' or 'y', not {variable}")
```
added `# pragma: no cover`

drbenvincent commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+        """Plots the model fit (posterior predictive) of the background products."""
+
+        if variable not in ["mu", "y"]:
+            raise ValueError(f"variable must be either 'mu' or 'y', not {variable}")
```
added `# pragma: no cover`

drbenvincent commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+            )
+
+            # likelihood
+            pm.Normal(
```
Good point. I've switched to a normal truncated at zero as a simple approach.

drbenvincent commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+                "intercept",
+                mu=pm.math.mean(background_sales[:treatment_time], axis=0),
+                sigma=np.std(background_sales[:treatment_time], axis=0),
+                # sigma=20,
```
Have removed commented out code

drbenvincent commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+                *list(background_sales.columns),
+                "new",
+            ],  # for non-saturated market only
+            # "all_sources": list(background_sales.columns)
```
Have removed commented out code

drbenvincent commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+            )
+        return model
+
+    @property
```
Done. And dealt with some consequences of this function having a parameter - i.e. passing that parameter around wherever it is called.

drbenvincent commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,403 @@
+HDI_ALPHA = 0.5
+
+
+class MVITS:
```
I think this is a good idea. Might as well get it integrated in to pymc-examples nicely right from the start.

wd60622 replied:
I agree plus it gives you save & load. If you need any support with this, let me know. I think I finally have it down pretty well

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,445 @@
+                    self.idata,
+                    var_names=["mu", "y"],
+                    random_seed=self.rng,
+                    predictions=True,
```
If you want to expose the counterfactual to the users and find `model.idata.predictions` confusing, then we can explore a property to expose it via a different name. What's your pulse on the usage and users?

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,445 @@
+                .transpose(..., "time")
+                .sel(background_product=background_product),
+                fill_kwargs={
+                    "alpha": HDI_ALPHA,
```
Shall we expose this to the user?

wd60622 commented on pyproject.toml:
```diff
@@ -79,6 +79,7 @@ packages = [
     "pymc_marketing.mmm.components",
     "pymc_marketing.clv",
     "pymc_marketing.clv.models",
+    "pymc_marketing.product_incrementality"
```
I will make a new issue for this. Here: #995

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,445 @@
+            self.market_saturated,
+            treatment_time=self.treatment_time,
+        )
+        self.sample_prior_predictive()
+        self.fit()
+        self.sample_posterior_predictive()
+        self.calculate_counterfactual()
```
The other classes have separation between initialization and sampling. Should be continue with that here? It also might make the testing easier. WDYT?

wd60622 reviewed (commented):
Some additional comment and ideas

drbenvincent commented:
Note to self... Use [`test_model_builder.py`](https://github.com/pymc-labs/pymc-marketing/blob/incrementality/tests/test_model_builder.py) to guide the mission of making the `MVITS` class inherit from `ModelBuilder`. In that file is a test class with the minimum set of abstract methods implemented. It also has tests, and I'll need to add such tests.

wd60622 commented:
Can you add an item to the PULL_REQUEST_TEMPLATE.md for this module

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,445 @@
+
+    def calculate_counterfactual(self):
+        """Calculate the counterfactual scenario of never releasing the new product."""
+        zero_sales = np.zeros(self.data[self.treatment_sales].shape, dtype=np.int32)
```
We might have to get the dtype from the input data here

wd60622 commented on README.md:
```diff
@@ -1,6 +1,6 @@
 <div align="center">
 
-![PyMC-Marketing Logo](docs/source/_static/marketing-logo-light.jpg)
+![PyMC-Marketing Logo](https://github.com/pymc-labs/pymc-marketing/blob/main/docs/source/_static/marketing-logo-light.jpg)
```
```suggestion
![PyMC-Marketing Logo](docs/source/_static/marketing-logo-light.jpg)
```

juanitorduz commented:
Fantastic work @wd60622 ! 🚀 ! Let's push this one to the finish line and cut a release :D !

drbenvincent commented:
> Fantastic work @wd60622 ! 🚀 ! Let's push this one to the finish line and cut a release :D !

Just a note of caution on merging this - we only really get one chance to release this as a new module, and I think the framing of the new module is important. At the moment all the guide and how-to docs are all framed as **product incrementality**. But there should be some discussion of whether this is the best thing. For example, it could be more pithy and punch if we called it something like an **uplift** module. Though some have proposed something like **sales analytics** - though that might be too broad because MMM could fall under that banner. 

I'll try to engage people again in the Labs discord to try to reach a consensus.

drbenvincent commented:
I now officially don't like the term **product incrementality**. But please add your thoughts to some of the suggestions I've dropped into Discord.

wd60622 commented:
I am having a bug with the predictions in the model. They seem to be overwriting the posterior_predictive group. This can be seen in the charts that have the `plot_fit` being the same as the plots from the `plot_counterfactual` method calls. 


Here is a basic example:

```python
from pymc_marketing.product_incrementality import MVITS

drbenvincent commented:
So just flagging up that in the change over to using `ModelBuilder`, the results have changed. In the `mv_its_saturated.ipynb` the first trace plot used to look like this:

![before](https://github.com/user-attachments/assets/17d0e7ed-ad63-411a-bb6e-d8e9652e913e)

But now looks like this:

![now](https://github.com/user-attachments/assets/c05b098f-eb65-4b83-9458-c85c9910d58e)

Maybe just a change in the priors, but the results we have before "made sense".

wd60622 commented:
Made changes in the saturated notebook. Would have to add some comments as well. 

Need to make changes to other one as well. I will be afk for bit so feel free to take a stab @drbenvincent

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,573 @@
+    def __init__(
+        self,
+        existing_sales: list[str],
+        market_saturated: bool = True,
```
I find `market_saturated` reads a bit weird. What are your thoughts on `saturated_market` instead? ie `MVITS(..., saturated_market=True)`

drbenvincent replied:
I'd be happy with `saturated_market`

drbenvincent replied:
Done in https://github.com/pymc-labs/pymc-marketing/pull/803/commits/c96a56803a769e17ee77dcbc3d206b938634a83a

wd60622 commented on pymc_marketing/product_incrementality/mv_its.py:
```diff
@@ -0,0 +1,573 @@
+        return plot_product(data=data, ax=ax)
+
+
+def plot_product(data: pd.DataFrame, ax: plt.Axes | None = None) -> plt.Axes:
```
Total sales with many products is on a different scale. I think this should be a flag and expose it in the various plotting methods

drbenvincent replied:
Agree!

drbenvincent replied:
resolved by https://github.com/pymc-labs/pymc-marketing/pull/803/commits/704fc2577dff6e4c9e7d7637a62eb7123be5ec52

drbenvincent commented:
That last commit was a bit of a big one.
* changed the package name from `product_incrementality` to `cc`
* ran `make uml` and added new images into `CONTRIBUTING.md`

drbenvincent commented:
Not sure why the `streamlit` tag was auto added to this PR

wd60622 commented on docs/source/api/index.md:
```diff
@@ -13,7 +13,7 @@
   model_builder
   prior
-  product_incrementality
+  cc
```
Don't think this should be abbreviated

drbenvincent replied:
as in the module name should be `customer_choice`, changing `cc` to `customer_choice` globally?

wd60622 replied:
yes

drbenvincent commented:
Getting crazy mypy errors in the pre-commit. Pushed the last one with `--no-verify` because it's late on a Friday and it seems unrelated to what I did adding the `plot_total_sales` kwarg 😅

wd60622 commented:
All looks really clean @drbenvincent Nice work!
I will take a look later and can do those edits to docstrings, tests, etc when I get the chance

wd60622 commented on docs/source/getting_started/quickstart/customer_choice/index.md:
````diff
@@ -1,5 +1,41 @@
 
-TODO - add python code block below with minimal code to get started with customer choice modeling
+```python
+from pymc_marketing.customer_choice.mv_its import (
````
Can we import in customer_choice/__init__ so we can use
`from pymc_marketing.customer_choice import MVITS, ...`

drbenvincent replied:
done :)

wd60622 commented:
> @wd60622 All tests pass again now - aside from code coverage. Am I right in understanding that you might do some work on this PR? I probably need to put this aside until late next week until things free up a bit.

Nice. Yes, I take a look today

juanitorduz commented:
@drbenvincent what is the plan for this PR?

juanitorduz commented:
> I think it's very close to done. Would appreciate a review on any of the 'packaging' aspects and docs etc

@drbenvincent I will have the time to look into this one during this end-of-the-year break 🙏 . it would be great to have it merged by January for our next release ... let's see :)

juanitorduz reviewed (approved):
@wd60622 @drbenvincent  I resolved the conflicts of https://github.com/pymc-labs/pymc-marketing/pull/803/ and my suggestion is to merge it as it is becasue I think this is in a good state and we can work on faster iterations in the next weeks before the release. WDYT? I do not want this to be forgotten and we can warn users this is experimental ? WDYT?

juanitorduz commented:
I will create some issues for next iterations

juanitorduz commented:
The pre-commit error will be fixed after we merge https://github.com/pymc-labs/pymc-marketing/pull/1313

juanitorduz commented:
great work @drbenvincent ! Let's keep working on iterations

drbenvincent commented:
Great stuff. Thanks for all the review comments. I'm happy to write a social media post when the next release comes out, explaining that this is a minimal viable product which is in line for some improvements. 

I'm pretty much unable to do any work until Jan 6th though.

juanitorduz commented:
Sure! No rush! I thing is actually to post about it even in mid or late January! Enjoy your time off !
